File: 2021/2021W51/PreppindataW51.py
import pandas as pd

df = pd.read_csv("C:\\Data\\PreppinData\\2021W51 Input.csv", 
                     parse_dates=['Order Date'], date_parser=lambda x: pd.to_datetime(x, format='%d/%m/%Y'))
df['Order Date'] = df['Order Date'].dt.date
temp = df['OrderID'].str.split('-', n = 1, expand = True) 
df['Store'] = temp[0] 
df['OrderID'] = temp[1]
df['Returned'] = df['Return State'].apply(lambda x: int(pd.notna(x)))
df['Unit Price'] = df['Unit Price'].apply(lambda x: float(x[1:])) 
df['Sales'] = df['Unit Price']*df['Quantity']
df['Product'] = df['Product Name'].str.lower()

for dim in ['Store', 'Customer', 'Product']:
    df[f'{dim} First Order'] = df['Order Date'].groupby(df[dim]).transform('min')
    df[f'{dim}ID'] = df[[f'{dim} First Order',dim]].apply(tuple,axis=1).rank(method='dense',ascending=True).astype(int)

fact = df[['StoreID', 'CustomerID', 'OrderID', 'Order Date', 'ProductID', 'Returned', 'Quantity', 'Sales']].copy()

store = df[['StoreID', 'Store', 'Store First Order']].drop_duplicates().copy()
store = store.rename(columns={'Store First Order':'First Order'}).sort_values('StoreID')

customer = df.groupby(['CustomerID', 'Customer', 'Customer First Order'], as_index=False).agg({'OrderID':'nunique', 'Product':'count', 'Returned':'sum'})
customer['Return %'] = round(customer['Returned']/customer['Product'], 2)
customer = customer.rename(columns={'Customer First Order':'First Order', 'OrderID':'Number of Orders'})
customer = customer[['CustomerID', 'Customer', 'Return %', 'Number of Orders', 'First Order']].sort_values('CustomerID')

product = df[['ProductID', 'Category', 'Sub-Category', 'Product Name', 'Unit Price', 'Product First Order']].drop_duplicates().copy()
product = product.rename(columns={'Product First Order':'First Sold'}).sort_values('ProductID')

File: 2021/2021W51/preppindataW51.R
library(readr)
library(dplyr)
library(tidyr)

df <- read_csv("C:/Data/PreppinData/2021W51 Input.csv",
               col_types =cols(`Order Date` = col_date(format = "%d/%m/%Y"))) %>%
  separate(`OrderID`, c('Store','OrderID'),sep='-') %>%
  mutate('Returned' = as.integer(!is.na(`Return State`)),
         'Unit Price' = as.numeric(substring(`Unit Price`, 2)),
         'Sales' = `Unit Price`*`Quantity`) %>%
  group_by(`Store`) %>%
  mutate('Store First Order' = min(`Order Date`)) %>%
  group_by(`Customer`) %>%
  mutate('Customer First Order' = min(`Order Date`),
         'Number of Orders' = n_distinct(`OrderID`),
         'Return %' = round(sum(`Returned`)/n(),2)) %>%
  group_by(`Product Name`) %>%
  mutate('Product First Sold' = min(`Order Date`)) %>%
  ungroup() %>%
  mutate('StoreID' = dense_rank(interaction(`Store First Order`, `Store`, lex.order=T)),
         'CustomerID' = dense_rank(interaction(`Customer First Order`, `Customer`, lex.order=T)),
         'ProductID' = dense_rank(interaction(`Product First Sold`, tolower(`Product Name`), lex.order=T)))
  
fact <- df %>%
  select(c('StoreID', 'CustomerID', 'OrderID', 'Order Date', 'ProductID', 'Returned', 'Quantity', 'Sales'))

store <- df %>%
  select(c('StoreID', 'Store', 'Store First Order')) %>%
  distinct() %>%
  rename('First Order' = 'Store First Order') %>%
  arrange(`StoreID`)

customer <- df %>%
  select(c('CustomerID', 'Customer', 'Return %', 'Number of Orders', 'Customer First Order')) %>%
  distinct() %>%
  rename('First Order' = 'Customer First Order') %>%
  arrange(`CustomerID`)
  
product <- df %>%
  select(c('ProductID', 'Category', 'Sub-Category', 'Product Name', 'Unit Price', 'Product First Sold')) %>%
  distinct() %>%
  rename('First Sold' = 'Product First Sold') %>%
  arrange(`ProductID`)

View(fact)
View(store)
View(customer)
View(product)
